### DISCUSSION TITLE
Improvements to threads/Redis errors from 1.0 RC to 1.5?

### DISCUSSION THREAD

**Discussion by randaratceridian** [OrchardCMS/OrchardCore#12966] [General]
We have recently been having issues with some Orchard Core containers becoming unresponsive on our production site running in Azure on Linux. We opened up a ticket with Microsoft that did a detailed analysis pointing at several root causes.

- A high number of threads where the stack trace points to OrchardCore.Autoroute.Services.AutorouteStateDocument. These seem to start at OrchardCore.Infrastructure.dll!OrchardCore.Documents.DocumentManager1+<GetOrCreateImmutableAsync>. The number is high enough that we get thread exhaustion.
- Several Redis errors. This is the link they provided: The following article explains the Redis errors, the particular section of interest is ‘Are you seeing a high number of busyio or busyworker threads in the timeout exception?’ Please review the advice in this article to see if it may apply to your application. [Are you getting network or CPU bound? | StackExchange.Redis](https://stackexchange.github.io/StackExchange.Redis/Timeouts)

I have a lot more details I can provide, but there is a key item to bring up. We are currently running OC 1.0 RC and .Net 5.0. We are in the process of upgrading to OC 1.5 and .Net 6.0. Do you think it might help with the issues above? Are there changes that might help?

One of my devs looked through the release notes. His comments:

_I looked at the change logs on the Orchard Core Repo in Github, trying to find something that would help with the thread pool exhaustion problem._

_There is nothing specific about thread optimization or fixes, but there are tons of small optimizations that may help out, like fixing memory issues with AppSettings._

_Two changes that I think may indirectly relate to our problem are:_

_1.4: Use of sentinels for Redis (from a quick web search, it seems to be a way to use high-availability on non-clustered Redis)_

_1.3: Upgrade for the Linq.Async library and some other async issues_

**Comment by jtkech** [comment]
Yes there have been improvements and it is worth to migrate.

For example we now are more resilient when a Redis connection is broken (e.g. during maintenance windows), we fallback to the shared database to be in sync and then use the local memory cache until we can reconnect to Redis. The most sensitive thing may be to migrate from OC 1.0 to OC 1.5, take care to migrate in multiple steps and to backup your data.
